### ISSUE TITLE
Variable listeners (and other ways of getting data between unity and the story)

### ISSUE THREAD

**Issue by Poltroon** [inkle/ink#413]
In my planned game, on the Unity side there are a number of systems to do with character generation, battle, tech trees and so on. These systems get pretty heavy with some calculations I don't believe Ink is capable of, and they need access to pretty much all the variables to get and set them. So it makes sense to store the game state here.

But on the Ink side (which I'm using for missions and end of turn events), the story also needs access to  pretty much the same stuff. Some choices only appear if certain conditions are met. And almost everything can be changed by your choices in the story.

Does anyone have any thoughts about the possibility / best way of sharing all the needed data between the unity side and story side? 

Are variable listeners part of the answer?
How do they work?
The documentation says:

> You can register a delegate function to be called whenever a particular variable changes.
Which I don't understand.

 Do I simply put the following in the update method?:
```
_inkStory.ObserveVariable ("health", (string varName, object newValue) => {
    SetHealthInUI((int)newValue);
});
```
...or do I have to do something with functions on the Ink side?

**Comment by joningold** [comment]
There are loads of ways to communicate with the gamelayer.

- store all the values in ink, but call an external function on the
gamelayer when you want a calculation doing.
  -- the game-layer can access all ink variables, to look things up
  -- it can return a value
  -- external functions shouldn't modify values, since external functions
should only be used to test state (unless you know what you're doing; the
short of it is they might get called multiple times)
  -- so if you want to modify a  value, get the new value via an external
but apply it inside ink

- store all values in the game and push them into the ink / look them up
via external functions when needed
  -- so ink doesn't modify anything, just 'reflects' game state
  -- ink can still request variable changes; a good way to do this is to
have a line of content like >>> STRENGTH + 3  and parse it instead of
printing it to screen


- a mixture of the two, depending on context



Variable listeners watch for variables inside the ink changing value and
then act on them. This is good for, say, updating UI to reflect internal
ink values; or if you want to fire off complex consequences to some value
change somewhere.

cheers
jon

**Comment by Poltroon** [comment]
Thank you for the reply.

It's tricky because I would like both layers to be able to check and set almost anything, not for one to be just a reflection of the other.

As a blanket solution, would it be horrible to have variable observers always listening for every possible variable change within the story (in a large list), and get the game state in the unity layer updated that way?

**Comment by alpapan** [comment]
It would be awesome if we had an example of where this usually goes...! e.g would it go to under a story.canContinue (i'm looking at the Intercept) or in Enter?

```
You can register a delegate function to be called whenever a particular variable changes.  This can be useful to reflect the state of certain ink variables directly in the UI. For example:

_inkStory.ObserveVariable ("health", (string varName, object newValue) => {
    SetHealthInUI((int)newValue);
});
```
I'm also having problems following this without an example... sorry!:

```
The reason that the variable name is passed in is so that you can have a single observer function that observes multiple different variables.
```

**Comment by papataci** [comment]
Read with interest, but cannot figure out if i can read variable stored inside Unity Objects.

i create characters in Unity this way:

```
public class Character : MonoBehaviour
{
    public string characterName;
    public Sprite icon, fullFigure;
    public int health;
    public int damage;
    public int money;
    public List<string> inventoryList = new List<string>();
    public List<string> skillsSetList = new List<string>();
    public string mood;
    public enum Mood
    {
        noMood, pleased, neutral, upset, leaving
    }
}

```

how do i read the varibles stored in the Character?

can call functions from ink via binding, but can only write to the log:

```
public void GetCharacterHealth(string characterName)
    {
        Debug.Log("Health:"+GameObject.Find(characterName).GetComponent<Character>().health);
    }

```

can i put the Unity values in an Ink variable?

please help!

**Comment by exitshadow** [comment]
> can i put the Unity values in an Ink variable?

Yes, you can. You just have to declare your function with a return type, then register it. By example, I would like to link this function:

```
string GetPlayerName()
{
    return "Luis";
}
```

Now to actually get it in the story, I need to register it by accessing the registry and add it into the bindings. There are many, many overloads in Ink’s `Story.cs` that allow for register almost all Actions<T,..> and Func<TResult,...> (Funcs are Actions with a return type, which is what you want here).

So you write

```
InkStory story;
story.BindExternalFunction("getPlayerName", GetPlayerName);
```

If you were to create a big dictionary of functions and bind them all together, it can get annoying to replicate the overloads cases (or use dynamic typing in projects where it is affordable to do so). But if you directly call `story.BindExternalFunction` and you have less than 4 arguments and 1 return type, and that all your arguments are of types string, int, float and bool, you’re good to go.

Once in Ink, you just have to declare the external and assign your variable

```
EXTERNAL getPlayerName()
// then, this is not compulsory but it is recommended in the case your function
// accidentally wasn’t bound or had a problem during its execution or evaluation
// define a fallback that ink will execute if that error happens
== function (getPlayerName) ==
~ return "defaultName"



VAR playerName = ""

Guy: So, what is your name, young lad?
Player: Oh, let me ask my game.
Player: My name is { getPlayerName } // ink will evaluate the return type right in the curly brackets
```

For a case so simple though it is likely that you would have variables that reflect each other or variable observers that keep track of things like player name, etc. But let’s say you would like to have functions like

```
float CalculateRemainderOfLoanInterests(float loanAmount, float amountPaid, float endDate, float interest)
{
   float remainder;
    // do some important maths here
   return remainder;
}

float DistorsionInSpaceTime(string planetA, string planetB, bool useJulianCalendar)
{
   float distorsion;
   // call nasa api, wait and do some maths....
   return distorsion;
}
```

Then you have no other choice but an external. Now writing the second silly example I do wonder how does Ink respond if it awaits the return of an external function that needs to be asynchronous. I would assume it would be safer to have a function to first command the the calculation then await for execution in code and call the dialogue back when the result is ready.
